source/h2_frames.c

```diff
@@ -1228,7 +1226,8 @@ int aws_h2_encode_frame(
 
 void aws_h2_frame_encoder_set_setting_header_table_size(struct aws_h2_frame_encoder *encoder, uint32_t data) {
     /* Setting for dynamic table size changed from peer, we will update the dynamic table size when we encoder the next
-     * header block */
+     * header block. The spec requires us to update the table if the updated size is smaller than the current size, but
+     * we can still update it even if not */
     aws_hpack_set_max_table_size(encoder->hpack, data);
     aws_hpack_set_protocol_max_size_setting(encoder->hpack, data);
```

Comment:
I just spent like an hour trying to make sense of whether or not we're doing the right thing with this hpack resize stuff. MAN this stuff is hard to understand. We just have a lot of variables with similar names, and the 1 hpack class that has both encoding and decoding logic. Can I make some suggestions?

1) Don't have `aws_h2_frame_encoder_set_setting_header_table_size()` call `aws_hpack_set_protocol_max_size_setting()` at all. Turns out that function has no effect on an aws_hpack_context that is being used for encoding.

2) rename ~aws_hpack_set_max_table_size()~ ->
```
/* Call this after receiving SETTINGS_HEADER_TABLE_SIZE from peer and sending the ACK.
 * The hpack-encoder remembers all size updates, and makes sure to encode the proper
 * number of Dynamic Table Size Updates the next time a header block is sent. */
AWS_HTTP_API
void aws_hpack_encoder_update_max_table_size(struct aws_hpack_context *context, uint32_t new_max_size);
```

3) rename ~aws_hpack_set_protocol_max_size_setting()~ ->
```
/* Call this after sending SETTINGS_HEADER_TABLE_SIZE and receiving ACK from the peer.
 * The hpack-decoder remembers all size updates, and makes sure that the peer
 * sends the appropriate Dynamic Table Size Updates in the next header block we receive. */
AWS_HTTP_API
void aws_hpack_decoder_update_max_table_size(struct aws_hpack_context *context, uint32_t new_max_size);
```

4) I still don't think `aws_hpack_set_protocol_max_size_setting()` aka `aws_hpack_decoder_update_max_table_size()` is doing the right thing. In the same way the encoder tracks a `smallest` and `last` size, I think we need to do similar tracking for the decoder so we can properly check that the peer sends the right number of updates?

man, this is why we need to split ~aws_hpack_context~ into `aws_hpack_encoder` and `aws_hpack_decoder`. It's too confusing having everyone's variables and functions all mixed up together